lib/compiler/src/unity_algorithm.cc

```diff
@@ -0,0 +1,472 @@
+            widen<SerialParallelDecomposition>(t.children[1])};
+  }
+  T decompn1 = t;
+  decompn1.children.pop_back();
+  return {decompn1, widen<SerialParallelDecomposition>(t.children.back())};
+}
+
+GraphSplit
+    get_graph_split(SerialParallelDecomposition const &pre_decomposition,
+                    SerialParallelDecomposition const &post_decomposition) {
+  return {get_nodes(pre_decomposition), get_nodes(post_decomposition)};
+}
+
+std::pair<SubParallelComputationGraph, SubParallelComputationGraph>
+    apply_split(SubParallelComputationGraph const &g, GraphSplit const &split) {
+  OpenMultiDiGraphView g1 =
+      unsafe_view_as_subgraph(OpenMultiDiGraph(g), split.first);
+  OpenMultiDiGraphView g2 =
+      unsafe_view_as_subgraph(OpenMultiDiGraph(g), split.second);
+
+  if (get_cut(OpenMultiDiGraph(g), split).size() > 0) {
+    // Sequential split
+    if (get_open_sinks(g1).size() <= get_open_sources(g2).size()) {
+      // get_open_sinks(*g1).size() should be 1 in perfect sp graphs
+      return {get_subgraph(g,
+                           split.first,
+                           InputSettings::INCLUDE_INPUTS,
+                           OutputSettings::EXCLUDE_OUTPUTS),
+              get_subgraph(g,
+                           split.second,
+                           InputSettings::INCLUDE_INPUTS,
+                           OutputSettings::INCLUDE_OUTPUTS)};
+    } else {
+      return {get_subgraph(g,
+                           split.first,
+                           InputSettings::INCLUDE_INPUTS,
+                           OutputSettings::INCLUDE_OUTPUTS),
+              get_subgraph(g,
+                           split.second,
+                           InputSettings::EXCLUDE_INPUTS,
+                           OutputSettings::INCLUDE_OUTPUTS)};
+    }
+  } else {
+    // Parallel split
+    return {get_subgraph(g,
+                         split.first,
+                         InputSettings::INCLUDE_INPUTS,
+                         OutputSettings::INCLUDE_OUTPUTS),
+            get_subgraph(g,
+                         split.second,
+                         InputSettings::INCLUDE_INPUTS,
+                         OutputSettings::INCLUDE_OUTPUTS)};
+  }
+}
+
+std::vector<std::pair<MachineSpecification, MachineSpecification>>
+    get_resource_split(MachineSpecification const &resource) {
+  std::vector<std::pair<MachineSpecification, MachineSpecification>> result;
+  for (int i = 1; i < resource.num_nodes; ++i) {
+    MachineSpecification sub_resource1 = resource, sub_resource2 = resource;
+    sub_resource1.num_nodes = i;
+    sub_resource2.num_nodes = resource.num_nodes - i;
+    result.push_back(std::make_pair(sub_resource1, sub_resource2));
+  }
+  return result;
+}
+
+Strategy::Strategy(float runtime,
+                   std::unordered_map<Node, MachineView> machine_views)
+    : runtime(runtime), machine_views(machine_views) {}
+
+bool Strategy::operator<(Strategy const &s) const {
+  return runtime < s.runtime;
```

Comment:
I think I might prefer a separate comparison functor here similar to `GraphCostComparator` from the old code in case this ordering is important, since many orders in FF code at the moment are just arbitrary (though well-defined) ones that exist to enable `std::map`, etc.